src/HIE/Bios/Config.hs

```diff
@@ -96,38 +152,61 @@ parseCradleType o
     | Just val  <- Map.lookup "other" o = Other <$> parseJSON val <*> pure val
 parseCradleType o = fail $ "Unknown cradle type: " ++ show o
 
-parseStackOrCabal
-  :: (Maybe String -> CradleType a)
-  -> ([(FilePath, String)] -> CradleType a)
+parseSingleOrMultiple
+  :: Monoid x
+  => (x -> CradleType a)
+  -> (x -> [(FilePath, x)] -> CradleType a)
+  -> (Map.HashMap T.Text Value -> Parser x)
   -> Value
   -> Parser (CradleType a)
-parseStackOrCabal singleConstructor _ (Object x)
-  | Map.size x == 1, Just (String stackComponent) <- Map.lookup "component" x
-  = return $ singleConstructor $ Just $ T.unpack stackComponent
-  | Map.null x
-  = return $ singleConstructor Nothing
-  | otherwise
-  = fail "Not a valid Configuration type, following keys are allowed: component"
-parseStackOrCabal _ multiConstructor (Array x) = do
-  let parseOne e
+parseSingleOrMultiple single multiple parse = doParse where
```

Comment:
Interestingly, this allows this configuration: 
``` yaml
cradle:
  cabal:
    components:
      - path: "./src"
        component: "lib:hie-bios"
      - path: "./"
        component: "lib:hie-bios"
    components:
      - path: "./src"
        component: "exe:hie-bios"
      - path: "./"
        component: "exe:hie-bios"
``` 
with this result:
``` haskell 
Config
  { cradle =
      CradleConfig
        { cradleDependencies = []
        , cradleType =
            CabalMulti
              { defaultCabal = Cabal { component = Nothing }
              , subCabalComponents =
                  [ ( "./src" , Cabal { component = Just "exe:hie-bios" } )
                  , ( "./" , Cabal { component = Just "exe:hie-bios" } )
                  ]
              }
        }
  }
```

Reply:
That's a good catch, didn't think about multiple `components` entries. I'll fix this before the PR is merged.

Reply:
Thinking about this a little bit more, isn't a duplicate key (`components`) an error according to the [yaml spec](https://yaml.org/spec/1.1/#id861060)?

> Mapping nodes are somewhat tricky because their keys are unordered and must be unique.

In that case, I suppose [this PR](https://github.com/snoyberg/yaml/pull/146/files) is relevant and https://github.com/mpickering/hie-bios/pull/230/files#diff-0478bef4574ee638b0c136cccc31479aR288 should use [`decodeFileWithWarnings`](https://www.stackage.org/haddock/lts-16.10/yaml-0.11.4.0/Data-Yaml.html#v:decodeFileWithWarnings)

Reply:
I added a test case that's supposed to fail on the config with duplicate `components` key.